torchtune/datasets/_concat.py

```diff
@@ -0,0 +1,134 @@
+        seed (int): seed for shuffling. Default is 29.
+        train_split_name (str): name of the split to use. Default is None, which will use the full dataset.
+        **load_dataset_kwargs (Dict[str, Any]): additional keyword arguments to pass to `load_dataset`.
+    """
+
+    def __init__(
+        self,
+        tokenizer: Tokenizer,
+        source: str,
+        text_column: str,
+        max_seq_len: int,
+        max_rows: int = None,
+        is_local: bool = False,
+        shuffle_before_packing: bool = False,
+        seed: int = 29,
+        train_split_name: str = None,
+        **load_dataset_kwargs: Dict[str, Any],
+    ):
+
+        assert tokenizer is not None, "tokenizer must be provided"
+        assert source is not None, "source must be provided"
+        assert text_column is not None, "text_column must be provided"
+        assert max_seq_len is not None, "max_seq_len must be provided"
+
+        if is_local:
+            dataset = load_from_disk(source, **load_dataset_kwargs)
+        else:
+            dataset = load_dataset(source, **load_dataset_kwargs)
+
+        if train_split_name is not None:
+            dataset = dataset[train_split_name]
+
+        if shuffle_before_packing:
+            dataset = dataset.shuffle(seed=seed)
+
+        # where final samples will be held
+        self.samples = []
+
+        # buffer to hold samples until they are long enough to be added to self.samples
+        buffer = []
+
+        def tokenize(sample):
+            tokenized = tokenizer.encode(
+                sample[text_column].strip(),
+                add_bos=True,
+                add_eos=True,
+            )
+            return {"input_ids": tokenized}
+
+        pbar_total = None
+        if max_rows is not None:
+            pbar_total = max_rows
+        # trying to get length of streaming dataset will throw an error
+        elif (
+            "streaming" in load_dataset_kwargs
+            and load_dataset_kwargs["streaming"] is not True
+        ):
+            pbar_total = len(dataset)
+
+        # create max_rows number of packed samples. If max_rows is None, create as many as possible
+        pbar = tqdm(
+            dataset, total=pbar_total, desc="Packing dataset", dynamic_ncols=True
+        )
+        for sample in dataset:
+            if max_rows is not None:
+                if len(self.samples) >= max_rows:
+                    break
+
+            sample = tokenize(sample)
+
+            buffer = buffer + sample["input_ids"]
+            while len(buffer) >= max_seq_len + 1:
+                if max_rows is not None:
+                    if len(self.samples) >= max_rows:
+                        break
+
+                self.samples.append(buffer[:max_seq_len])
+                buffer = buffer[max_seq_len:]
+
+                if max_rows is not None:
+                    pbar.update(1)
+
+            if max_rows is None:
+                pbar.update(1)
+
+    def __getitem__(self, idx):
```

Comment:
ideally we shouldn't have to use the `__getitem__` method, for streaming/packing an Iterable dataset would be more suitable than our current map-style dataset. Do you think that would make the implementation easier?